android/app/src/main/java/org/uwaterloo/subletr/pages/listingdetails/ListingDetailsPageViewModel.kt

```diff
@@ -0,0 +1,65 @@
+package org.uwaterloo.subletr.pages.listingdetails
+
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.util.Base64
+import androidx.lifecycle.SavedStateHandle
+import androidx.lifecycle.ViewModel
+import dagger.hilt.android.lifecycle.HiltViewModel
+import io.reactivex.rxjava3.core.Observable
+import io.reactivex.rxjava3.schedulers.Schedulers
+import io.reactivex.rxjava3.subjects.BehaviorSubject
+import kotlinx.coroutines.runBlocking
+import org.uwaterloo.subletr.api.apis.DefaultApi
+import org.uwaterloo.subletr.api.models.ListingDetails
+import org.uwaterloo.subletr.services.INavigationService
+import java.util.Optional
+import javax.inject.Inject
+import kotlin.jvm.optionals.getOrNull
+
+@HiltViewModel
+class ListingDetailsPageViewModel @Inject constructor(
+	private val defaultApi: DefaultApi,
+	savedStateHandle: SavedStateHandle,
+	val navigationService: INavigationService
+) : ViewModel() {
+
+	private val listingIdStream: BehaviorSubject<Int> =
+		BehaviorSubject.createDefault(checkNotNull(savedStateHandle["listingId"]).toString().toInt())
+	private val imageIdsStream: BehaviorSubject<List<String>> = BehaviorSubject.createDefault(emptyList())
+	val favouritedStream: BehaviorSubject<Boolean> = BehaviorSubject.createDefault(false)
+	private val listingDetailsStream: Observable<Optional<ListingDetails>> = listingIdStream.map {
+		runBlocking {
+			val listing = defaultApi.listingsDetails(it)
+			favouritedStream.onNext(listing.favourited)
+			imageIdsStream.onNext(listing.details.imgIds)
+			Optional.of(listing.details)
+		}
+	}
+		.subscribeOn(Schedulers.io())
+	private val imagesStream: Observable<List<Bitmap>> = imageIdsStream.map {
+		runBlocking {
+			it.map { id ->
+				val encodedImage = defaultApi.listingsImagesGet(id)
+				val imageBytes = Base64.decode(encodedImage, Base64.DEFAULT)
+				BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size)
```

Comment:
Also, maybe we should make a global function for this, like `fun String.base64ToBitmap() {...}`

Reply:
yeah made a utils package instead with `ImageUtils`